tree/20_cfe_basics/package_lib.cf

```diff
@@ -53,10 +53,17 @@ bundle agent ncf_package(name, version, architecture, provider, state, options)
       "version_description" string => "in version ${version}";
     
     # Log message
-    !(!state_present.version_latest)::
-      "message" string => "${state_description} of package ${name}${architecture_description}${version_description}";
-    !state_present.version_latest::
-      "message" string => "${state_description} of package ${name}${architecture_description}${version_description} is not supported and";
+    pass3::
+      "message" string => "If you tried to install a virtual package, please use in place the concrete package you want to install.${const.endl}${state_description} of package ${name}${architecture_description}${version_description}",
+        ifvarclass => "${old_class_prefix}_not_ok|${class_prefix}_not_ok";
+
+    pass3.!(!state_present.version_latest)::
+      "message" string => "${state_description} of package ${name}${architecture_description}${version_description}",
+        ifvarclass => "!(${old_class_prefix}_not_ok|${class_prefix}_not_ok)";
+
+    pass3.!state_present.version_latest::
+      "message" string => "${state_description} of package ${name}${architecture_description}${version_description} is not supported and",
```

Comment:
I think this can never happen as `!state_present.version_latest` will always fail.